src/core/segment_reader.rs

```diff
@@ -393,3 +393,57 @@ impl<'a> Iterator for SegmentReaderAliveDocsIterator<'a> {
         result
     }
 }
+
+#[cfg(test)]
+mod test {
+    use core::{Index, SegmentReader};
+    use schema::*;
+
+    #[test]
+    fn test_iterator() {
+        let mut schema_builder = SchemaBuilder::new();
+        schema_builder.add_text_field("name", TEXT);
+        let schema = schema_builder.build();
+        let index = Index::create_in_ram(schema.clone());
+        let name = schema.get_field("name").unwrap();
+
+        {
+            let mut index_writer = index.writer(50_000_000).unwrap();
+            let mut doc1 = Document::default();
+            doc1.add_text(name, "tantivy");
+            index_writer.add_document(doc1);
+
+            let mut doc2 = Document::default();
+            doc2.add_text(name, "horse");
+            index_writer.add_document(doc2);
+
+            // we should now have one segment with two docs
+            index_writer.commit().unwrap();
+        }
+
+        {
+            let mut index_writer2 = index.writer(50_000_000).unwrap();
+            index_writer2.delete_term(Term::from_field_text(name, "horse"));
+
+            // ok, now we should have a deleted doc
+            index_writer2.commit().unwrap();
+        }
+
+        let segments = index.searchable_segments().unwrap();
+        assert_eq!(segments.len(), 2);
```

Comment:
It is actually possible to end up with 1 segments with this setting.

Reply:
ok. i'll write something more robust. :) that was from me sniffing around and not really applicable to the test. :)

Reply:
It should be quite deterministic with the 1 thread setting.

Reply:
Ok so to sum it up, as it might help you understanding the behavior.

When you use several threads, each thread is building its own little segment by picking off docs from the same queue. When you call commit all of the workers who have processed at least one document flush their work to disk and create one segment.

Also if you add a lot of document without committing, the worker have a RAM budget and will start flushing to disk once their RAM budget is exhausted.